### DISCUSSION TITLE
Extracting x509 subject information to user profile

### DISCUSSION THREAD

**Discussion by pstatham-exp** [keycloak/keycloak#30513] [Q&A]
I've got a basic Docker container running Keycloak and using some self signed certificates to enable x509 authentication.

I've followed instructions from:

* https://www.keycloak.org/server/containers
* https://www.keycloak.org/server/enabletls
* https://www.keycloak.org/docs/latest/server_admin/#_x509
* https://github.com/keycloak/keycloak/issues/15481
* https://www.youtube.com/watch?v=yq1hzNs1JQU

After following the above, I am successful in getting my secured app to require x509 certificates in order to access the application.

However, when using the Javascript Adapter, I've noticed that the returned authentication instance includes no information about the x509 that was used; any information about the user comes from the identified user profile (which is done by setting the x509 authentication configuration to use Subject Email Address and Username and Email lookup).

Does Keycloak have the capability to "store" or "map" authenticated certificate attributes to the authenticated user profile?

**Comment by sam-hayes-org** [comment]
I have exactly the same problem. I need to identify a specific user who executes such client authentication by the `emailAddress`  within the certificate provided to Keycloak during authentication. 

The `emailAddress` from the certificate's subject is NOT automatically mapped to the `email` scope within the returned access token.   This scope holds the email address of the Keycloak service account user that was automatically created when generating the client in Keycloak.

I am really not sure if any certificate subject information can be mapped to any scope at all.

**Comment by dteleguin** [comment]
@pstatham-exp @sam-hayes-org you can always use Keycloak's [X509ClientCertificateLookup](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/services/x509/X509ClientCertificateLookup.java) to access the client certificate from your custom authenticators, mappers etc.

**Reply by pstatham-exp** [reply]
@dteleguin We are using the JS Adapter and have our Keycloak instance setup to do X509 authentication. There is information available in the certificate a user authenticates with (org/email/etc.) that we would like to get back after authentication.

**Reply by dteleguin** [reply]
This is exactly what client mappers are for. If you wish to extract arbitrary data from the x509 certificate, you will need to code a custom mapper (in Java) and perform a certificate lookup from it.
